src/npctalk.cpp

```diff
@@ -4491,6 +4491,22 @@ dynamic_line_t::dynamic_line_t( JsonObject jo )
             const bool wearing = d.alpha->is_wearing( item_id );
             return ( wearing ? yes : no )( d );
         };
+    } else if( jo.has_member( "u_has_trait" ) ) {
+        std::vector<std::string> traits_to_check;
+        JsonArray jarr =  jo.get_array("u_has_trait");
+        while (jarr.has_more()) {
+            traits_to_check.push_back( jarr.next_string() );
+        }
```

Comment:
Those 5 lines (ending here) can be shortened to
```C++
const std::vector<std::string> traits_to_check = jo.get_string_array( "u_has_trait" );
```

----

Alternatively, you could use `get_tags` (which returns a set, not a vector)
```C++
const std::set<std::string> traits_to_check = jo.get_tags( "u_has_trait" );
```
Your code below works well with a set, no changes needed there.

If the JSON input contains a string as value for "u_has_trait" and not an array, `get_string_array` will fail and parsing will stop. `get_tags` handles this just fine.

This would work with `get_tags`, but not with `get_array` or `get_string_array`:
```JSON
{
    "u_has_trait": "QUICK",
    ...
}
```

----

Should the member be named "u_has_any_trait"? Giving `"u_has_trait":["QUICK","FLEET"]`, I would expect it to choose the "yes" line if the character has *both* traits.